### DISCUSSION TITLE
What is the best way to access serial ports in a tauri app?

### DISCUSSION THREAD

**Discussion by tgirotto** [tauri-apps/tauri#2108] [Q&A]
I can think of two options:

1) Use an npm package like serialport, but I am not sure it's possible;
2) Spin up a backend server that has access to the serial port apis and use http to communicate with the tauri app.

Any advice?

**Comment by nothingismagick** [comment]
Use https://docs.rs/serialport/4.0.1/serialport/ and create a tauri plugin that offers a JS API.

**Reply by thadguidry** [reply]
Great idea, expand the Tauri ecosystem, create a plugin, so that the next person with the same problem can just use it!  And then the community itself can help you maintain the plugin, you wouldn't be alone!

**Comment by carlos22** [comment]
There seems to be a plugin available here: https://github.com/lzhida/tauri-plugin-serialport

**Reply by bastian2001** [reply]
Did you get it to work? I can only write to the serial port, but not read from it. I do not understand the difference between read and listen, and I don't get any data, even though I have verified the program on the MCU to be functional (just a basic Serial mirror).

It also seems weird that the read function has no callback function to do something when data was received, or a buffer to put the bytes in once received. It also only returns a void Promise, so there's nothing to be taken from the read function, if I understand it correctly.

**Comment by nathanbabcock** [comment]
Doing it in Rust is probably the safest bet as people have mentioned, however there is also a Web Serial API to do it purely browser-side:

https://developer.mozilla.org/en-US/docs/Web/API/Web_Serial_API

Looks to be supported in Edge (Webview), if WIndows is your target platform: https://caniuse.com/web-serial

**Reply by carlos22** [reply]
But I doubt it is usable from WebView2. As this is a feature with security implications I think they disable it for embedded views or at least you need to configure something to make it work. But I did not try it...

**Reply by hanckmann** [reply]
Currently I am also looking into this and I would like to do it in Rust.
Still, the same question is valid... what is the good way to do this?

**Reply by hansSchall** [reply]
I can confirm that WebSerial works in WebView2/Tauri.

It would be nice if there were a permission handler api in rust (like https://www.electronjs.org/docs/latest/tutorial/devices#web-serial-api), instead of showing the default Edge serial permission dialog.

**Comment by earthtosid** [comment]
[This guy did it](https://github.com/deid84/tauri-plugin-serialport)

**Comment by s00d** [comment]
my version for tauri 2 https://github.com/s00d/tauri-plugin-serialplugin

**Reply by mrchypark** [reply]
may this support ios?
